Web page test

renajohn opened the pull request.

Commit:
Add a custom capability to wait before executing a new action

Add a new capability "lockStep":boolean. When set, the remote webdriver server
will poll a server every second to decide when the next action can be send.
New session is never delayed.

When a session will be either terminated (QUIT) or closed (CLOSE), an HTTP
request is sent first to the endpoint, then the server will wait for the ready
flag to be true.

At the moment, the endpoint is hardcoded to localhost:8888.

Commit:
Explose hook endpoint as a java property "hookEndPoint"

kart commented on java/server/src/com/appdynamics/wpt/WptHookAwareHandler.java:
```diff
@@ -0,0 +1,73 @@
+ * @since 2/14/2015
+ */
+public class WptHookAwareHandler {
+  private WptHookClient wptHookClient;
```
Probably my mistake. But, this should be final.

renajohn replied:
Not sure it's possible. The `WptHookClient` constructor requires parameters provided to the `WptHookAwareHandler` constructor. I'm not aware of a way to have a final object initialized in a constructor. Is that possible in Java?

kart commented on java/server/src/com/appdynamics/wpt/WptHookAwareHandler.java:
```diff
@@ -0,0 +1,73 @@
+    try {
+      wptHookClient = new WptHookClient(url, log);
+    } catch (MalformedURLException e) {
+      e.printStackTrace();
```
Again, my mistake, I guess. We should do log.error here and other places wherever appropriate.

kart commented on java/server/src/com/appdynamics/wpt/WptHookAwareHandler.java:
```diff
@@ -0,0 +1,73 @@
+
+    if (command.getName().equals(DriverCommand.QUIT) || command.getName().equals(DriverCommand.CLOSE)) {
+      // we need to inform WPT that the session will terminate
+      wptHookClient.webdriverDone();
```
This was the bug I ran into. We should return here. Otherwise, we will go further and call waitUntilHookReady().

kart replied:
We should probably do a log.info here too just so that we know we dispatched a webdriver_done as part of QUIT.

kart commented on java/server/src/com/appdynamics/wpt/WptHookAwareHandler.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  public boolean waitIfNeeded(Command command) throws Exception {
+    boolean hasWaited = false;
```
This doesn't seem to be used anywhere. Do you think we need this?

renajohn replied:
`hasWaited` is used in the code, where you referring to `getWptHookClient`?

kart commented on java/server/src/com/appdynamics/wpt/WptHookAwareHandler.java:
```diff
@@ -0,0 +1,73 @@
+        if (wptHookClient.isHookReady()) {
+          break;
+        }
+        Thread.sleep(1000);
```
Since we are not doing it in the client anymore, how about we make this 100ms?

renajohn replied:
Sure, that makes sense.

kart commented on java/server/src/com/appdynamics/wpt/WptHookAwareHandler.java:
```diff
@@ -0,0 +1,73 @@
+        Thread.sleep(1000);
+      } catch (IOException e) {
+        log.severe("Unable to get response for 'is_hook_ready' request!");
+        throw new RuntimeException("Cannot reach WptHook", e);
```
We probably should throw a better exception here.

renajohn replied:
Let's fix that later on, I don't have a good idea on what to put here.

kart commented on java/server/src/org/openqa/selenium/remote/server/rest/ResultConfig.java:
```diff
@@ -62,6 +59,9 @@ public ResultConfig(
     this.handlerFactory = getHandlerFactory(handlerClazz);
+
+    String hookEndPoint = System.getProperty("hookEndPoint", "http://localhost:8888");
+    this.wptHookAwareHandler = new WptHookAwareHandler(log, hookEndPoint);
```
I am not sure about how the server is architected, but I wonder if we end up repeatedly creating new objects? If so, we should consider having a singleton or some such.

kart commented on java/client/src/org/openqa/selenium/remote/CapabilityType.java:
```diff
@@ -66,4 +68,13 @@
+   * @since 7/1/15
+   * Copyright 2015 AppDynamics inc. All rights reserved
+   */
+  final String LOCK_STEP = "lockStep";
```
LOCK_STEP seems too generic. How about LOCK_STEP_COMMAND_EXECUTION? Also, a brief overview of what it means would be helpful here. If not, at least a {@link} to WptHookAwareHandler would be good.

Commit:
Make WptHookAwareHandler a singleton initialized lazily

Lazy initialization is required since we have to pass the log object passed by argument to
the ResultConfig constructor.

Commit:
Address most PR comment + fix exception when hook not yet up and running

Commit:
Fix message order for end of session

Commit:
Rename lockStep to wptLockStep and add a comment

Commit:
Add firefox_profile_dir capability

This capability allows the driver to read the Firefox Profile from a specified
directory. This is an alternative to the firefox_profile capability, which reads
the entire profile directly.

Commit:
Merge branch 'WebPageTest' into firefox_profile

Commit:
Merge pull request #2 from Appdynamics/firefox_profile

Firefox profile directory

renajohn commented:
Closing PR since we will keep this branch separated from master.
